Del The World: Unit testing for hard deletes

LemonInTheDark opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Adds a unit test that runs at the very end of all other tests. It creates and then instantly destroys one of each type path. Some are blacklisted because of various reasons.

This won't catch anywhere near everything, but it'll help a lot with regressions, and catches a lot of runtimes too. The really nasty stuff is things that sleep as a part of initialize, since we don't init, wait, then qdel since it's a lot more prone to error

Oh and I went through and fixed all the errors I could, alongside adding descriptive errors on the stuff that's just gonna runtime in init.

This is gonna cause some unit test failures, but hopefully it'll just catch the simple stuff, which should be easy for people to fix.

I'm gonna add a label for fixing harddels, alongside a gbp boost of 5. I'd like people other then me to start at least somewhat caring about these

On the more technical side, @ninjanomnom asked me to add support for running tests on all objects of whatever type. Wanted to ask @Mothblocks for their opinion about it first tho.

## Why It's Good For The Game

\<Read the guide\>
[The Guide](https://github.com/LemonInTheDark/tgstation/blob/del-the-world/.github/HARDDEL_GUIDE.md)
\</Read the guide\>

This'll allow us to prevent hard delete regression, and it'll catch a good amount of general errors too since we're guaranteed to hit the init and destroy procs of nearly every typepath, which should help prevent runtime spam from say, spooky material not cleaning up properly or whatever.

~~Speaking of which, there's still some errors left on this pr, I'm hoping they're related to #59608, because otherwise I'm gonna have to actually fix them, which I don't relish.~~ They weren't, but I fixed them. Turns out connect loc is dumb, and pretty much the only solution was a `if(runtime) return` sorta deal.

Oh also I redid how tanks generated their parents list, it was causing warnings because they were doing it by relying on update_parents rather then using on_construct, which should really just be a mapload thing by the by

Lemme see, uh, this increases the awareness of hard deletes and prevents the bi yearly maintainer screed going through a list of recent harddels and just fixing whatever they can get their hands on.

This'll also catch all the bits of code that fail to return qdel/initialize hints, or do other naughty stuff. 

Seriously tho I just want atmos cpu time man is that so much to ask

LemonInTheDark committed:
Adds a del() unit test, it tests the integrity of our code's response to being destroyed, and prevents simple hard delete regressions

LemonInTheDark committed:
Fixes some runtimes sourced from code that expects init to pass in non standard input

LemonInTheDark committed:
Please stop deleting gas_mixtures, they're meant to free when they're no longer ref'd. If I catch you motherfuckers doing this shit in excited group handling you're gonna catch my hands

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into del-the-world

LemonInTheDark committed:
Fixes some annoying runtimes

LemonInTheDark committed:
Makes a lot of stuff safer for null inputs, more descriptive errors, also fixes some runtimes

LemonInTheDark committed:
Fixes an error in shuttle docker consoles, they used outdated logic for working with SSshuttles.stationary

LemonInTheDark committed:
Adds a lot more exceptions to the deletion unit test

LemonInTheDark committed:
More bespoke errors in initialize

LemonInTheDark committed:
Fixes some errors I came across in the process of doing this. Mostly spurious runtimes, a few behavior stuff

LemonInTheDark committed:
Updates the unit test to fill in more stuff

LemonInTheDark committed:
Fixes all uses of Destroy() that don't call parent

LemonInTheDark committed:
Make sure to clean up after yourself

LemonInTheDark committed:
Fixes an error in flag code

LemonInTheDark committed:
Adds more blacklisted types, gives you more control over reference tracking

LemonInTheDark committed:
Whoops lost this somehow

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into del-the-world

LemonInTheDark committed:
Fixes a potential hanging ref source

LemonInTheDark committed:
Makes the round properly end, fixes up some hanging refs in del_test, and adds one more thing to ignore, it's hip to fuck beams

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into del-the-world

LemonInTheDark committed:
Fucking mech

LemonInTheDark committed:
Fixes a monkey runtime, you can't find a qdeleted item

LemonInTheDark committed:
Makes a guide for hard deletes, inserts it into the contribution guidelines

LemonInTheDark committed:
Enables harddel ref searching as a part of the unit test, puts a time out at like 8 minutes, I'm hoping the github servers are powerful enough for this

LemonInTheDark committed:
Adds a label for hard deletes

LemonInTheDark commented:
I used this to generate a large portion of #59371, which is why it's not catching anything. Trust me on this, it took me like 5 days to fix all of em, there were like 200 erroring typepaths

LemonInTheDark committed:
Removes the hardlookup define

MrDoomBringer commented:
holy crap dude good work

Mothblocks commented:
man this is going to fail so sporadically its gonna be hillarious

we need it though

Mothblocks commented on .github/HARDDEL_GUIDE.md:
````diff
@@ -0,0 +1,263 @@
+
+You want your coders to be able to pass around objects without doing a full copy. So you'll store the pack of data somewhere in memory
+
+```c++
````
```suggestion
```dm
```

Why?

LemonInTheDark replied:
dm doesn't actually highlight. I used c++ cause it looked the best in my markdown previewer

Mothblocks commented on .github/HARDDEL_GUIDE.md:
````diff
@@ -0,0 +1,263 @@
+Then you them to be able to pass that object into say a proc, without doing a full copy. So you let them pass in the object's location in memory instead
+This is called passing something by reference
+
+```c++
````
```suggestion
```dm
```

Mothblocks commented on .github/HARDDEL_GUIDE.md:
````diff
@@ -0,0 +1,263 @@
+
+This of course means they can store that location in memory in another object's vars, or in a list, or whatever
+
+```c++
````
```suggestion
```dm
```

Mothblocks commented on .github/HARDDEL_GUIDE.md:
````diff
@@ -0,0 +1,263 @@
+
+So I guess you should scan all of memory for that reference?
+
+```c++
````
```suggestion
```dm
```

Mothblocks commented on .github/HARDDEL_GUIDE.md:
````diff
@@ -0,0 +1,263 @@
+
+First and simplest we have `Destroy()
+
+```c++
````
```suggestion
```dm
```

Mothblocks commented on .github/HARDDEL_GUIDE.md:
````diff
@@ -0,0 +1,263 @@
+
+Next, and slightly more complex, pairs of objects that reference each other
+
+```c++
````
```suggestion
```dm
```

Mothblocks commented on .github/HARDDEL_GUIDE.md:
````diff
@@ -0,0 +1,263 @@
+
+There's no need to clean these up, just make sure you account for it being null, since it'll return that if the object doesn't exist or has been queued for deletion
+
+```c++
````
```suggestion
```dm
```

Mothblocks commented on .github/HARDDEL_GUIDE.md:
````diff
@@ -0,0 +1,263 @@
+
+Here's an example
+
+```c++
````
```suggestion
```dm
```

Mothblocks reviewed (changes requested):
ill review this later

Mothblocks commented on code/modules/unit_tests/unit_test.dm:
```diff
@@ -28,6 +28,8 @@ GLOBAL_VAR(test_log)
 	var/focus = FALSE
 	var/succeeded = TRUE
+	//Do we have our own handling?
+	var/snowflake = FALSE
```
This is a dumb variable name, name it what it actually is. Just do `auto_run = TRUE`.

LemonInTheDark replied:
it seems like you want a priority system here instead, so let me nuke it totally

Mothblocks commented on code/modules/unit_tests/unit_test.dm:
```diff
@@ -78,49 +80,59 @@ GLOBAL_VAR(test_log)
 		CHECK_TICK
 
+	//Fuck you, fight me
+	RunUnitTest(/datum/unit_test/del_test, test_results)
```
This is dumb, just have a priority var that you sort a list of tests by.

LemonInTheDark replied:
sure

LemonInTheDark committed:
Makes unit tests priority based, esures that the del test won't cause failures if checking del results takes too long

LemonInTheDark committed:
Nicifies the doc

LemonInTheDark committed:
initial

Rohesie commented on .github/gbp.toml:
```diff
@@ -9,6 +9,7 @@ reset_label = "GBP: Reset"
 "Feedback" = 2
 "Fix" = 3
 "Grammar and Formatting" = 1
+"Hard Deletes" = 5
```
Performance is 12, but Hard Deletes, the major source of perf improvements, is only 5?
Outrageous.

LemonInTheDark replied:
performance is 12? damn
was just kinda a ballpark number honestly

Rohesie replied:
Perf good. Resolving hard dels good. Give it GBPs, we really want this kind of PR.

Mothblocks replied:
performance hsould just come with hard deletes

LemonInTheDark replied:
You think? Part of me hopes that seeing a fancy label will encourage people to look into them

Mothblocks replied:
i want them both

Mothblocks replied:
but any pr with hard deletes should also have performance is what i mean

LemonInTheDark replied:
want me to make it a bit less valuable then? or should we just shower em in it

ninjanomnom replied:
> 
> 
> but any pr with hard deletes should also have performance is what i mean

Performance is for prs with information about a specific performance increase and change meant specifically to improve performance. A plain hard del removal is a code cleanliness and stability thing that doesn't matter for performance most of the time.

Now if you fix a particularly big one that's different.

LemonInTheDark committed:
I may be stupid

LemonInTheDark committed:
Roh made a good point, let's bring this to parity with performance. Call it by a different name so people have an incentive to look into it

LemonInTheDark committed:
Looks like I broke picking things up, not my worst failure I'll admit

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into del-the-world

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into del-the-world

LemonInTheDark commented on .github/HARDDEL_GUIDE.md:
````diff
@@ -0,0 +1,265 @@
+    var/name = "some shit"
+```
+
+Then you them to be able to pass that object into say a proc, without doing a full copy. So you let them pass in the object's location in memory instead
````
Then you them (Thank you)

LemonInTheDark committed:
Then you them

Thank you steel

Co-authored-by: SteelSlayer <42044220+SteelSlayer@users.noreply.github.com>

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into del-the-world

LemonInTheDark committed:
Merge branch 'del-the-world' of https://github.com/LemonInTheDark/tgstation into del-the-world

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into del-the-world

Rohesie commented:
What's the current state of the PR?

LemonInTheDark commented:
> 
> 
> What's the current state of the PR

The unit test is failing for 3 reasons. A harddel and a signal logic error, which I'll pr a fix for later.
The final reason is an issue with connect_loc_behalf and moved(), stemming from space turfs occasionally sleeping on entered(). I'm working on it with kyler at the moment

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into del-the-world

LemonInTheDark commented:
Taking my two weeks paid vacation, I'll reopen this if I survive

LemonInTheDark committed:
temp

LemonInTheDark commented:
I appear to be alive, at least for now
Knock on wood

LemonInTheDark committed:
Merge remote-tracking branch 'tgstation/master' into del-the-world

LemonInTheDark committed:
Fixes some runtimes/harddels, cleans up some unneeded procs, some other stuff

LemonInTheDark committed:
Fixes some annoying runtimes

LemonInTheDark committed:
WAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA

LemonInTheDark committed:
Print to test

LemonInTheDark committed:
ful

LemonInTheDark committed:
Removes the initialize sanity test, I've implemented it seperately sans the not fully initing thing

LemonInTheDark committed:
Fixes a potential infinite loop caused by the area bound component, adds the component injector to the blacklist

LemonInTheDark committed:
Converts tram center refs to the global var, if this needs to be patched up that can be done later. Oh and I fixed a hanging ref in the global var

LemonInTheDark committed:
Smallest -> Largest

LemonInTheDark committed:
Blacklists one more thing

LemonInTheDark committed:
/Fixes/ the connect loc bug

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into del-the-world

LemonInTheDark committed:
JHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH

LemonInTheDark committed:
I hate callbacks

LemonInTheDark committed:
renames del_test to create_and_destroy

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into del-the-world

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into del-the-world

LemonInTheDark committed:
Fixes up some runtimes, one that assumed that a limb would always have a owner, the other that assumed that
ovens would not react to objects leaving post oven Destroy, which violates the logic around how objects moving
works

Why do they call it an oven when you of in the hot food cold in

Mothblocks commented on code/__DEFINES/traits.dm:
```diff
@@ -582,7 +582,7 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 
 ///generic atom traits
-/// Trait from [/datum/component/rust]. Its rusty and should be applying a special overlay to denote this.
+/// Trait from [/datum/element/rust]. Its rusty and should be applying a special overlay to denote this.
```
Why is this diff showing? Did you mess up a merge?

LemonInTheDark replied:
no clue man, I'll do a check to see if I missed anything tho

LemonInTheDark replied:
Turns out the guy who made rust into an element never actually updated this. So I'm right master is wrong?

Mothblocks commented on code/game/machinery/computer/tram_controls.dm:
```diff
@@ -7,7 +7,6 @@
 	flags_1 = NODECONSTRUCT_1
 	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF
 
-	var/obj/structure/industrial_lift/tram/tram_part
```
This seems like an unnecessary change that makes it harder to iterate on in the future. Are you sure you need this?

LemonInTheDark replied:
I did this because it was already kinda tied to how things work currently. I'm up for converting it to weakrefs, will do that tomorrow

LemonInTheDark replied:
done

Mothblocks reviewed (changes requested):
Last comment

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into del-the-world

LemonInTheDark committed:
Makes tanks properly use on_construct, removes a redundant update_parents from the merge proc. This logic is
dumb as fuck, and honestly on_construct should just be mapload, but hhhh

This'll fix all the warnings I think

LemonInTheDark committed:
Wooooot

LemonInTheDark committed:
Converts spammed global vars into weakrefs

Mothblocks commented on code/game/machinery/computer/tram_controls.dm:
```diff
@@ -9,6 +9,8 @@
 
 	light_color = LIGHT_COLOR_GREEN
 
+	var/datum/weakref/tram_ref
```
Document this because it's used outside of just tram_controls (in the component)

Last request

LemonInTheDark replied:
ah excuse me

LemonInTheDark committed:
doc

Cyberboss commented:
Bless
